flesh out workers concept doc a little (#390)
temporalio/documentation: docs/concepts/workers.md

sidebar_label: Workers
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

A Worker is a service that:

- hosts (registers) [Workflows](/docs/concepts/workflows) and [Activities](/docs/concepts/activities)
- listens to [Activity Task Queues](/docs/concepts/task-queues) by long polling

Workers should have access to the resources needed to perform the activities they implement, for example:

- network access for external API calls
- credentials for infrastructure provisioning
- specialized GPUs for machine learning tasks

Since workers are subscribed to task queues they can implement, it is straightforward to do [Activity Task Routing](/docs/concepts/activities/#activity-task-routing-through-task-queues). 

> Note: if you need to process stateful work sequentially on the same machine, the Go SDK also offers a [Sessions API](https://docs.temporal.io/docs/go/sessions/).

## Example Worker Code

> Just for illustration purposes:

Author: i'm not REALLY sure about this one, but i thought it'd be nice - let me know what you think! happy to remove.

Author: ```suggestion
## Example Worker Code

Here is some sample code to help you understand what a Worker does (as described above). Please refer to [each SDK's documentation](/docs/application-development) for a fuller description.
```

  const worker = await Worker.create({ workDir: __dirname, taskQueue: 'tutorial' });
  // Start accepting tasks on the `tutorial` queue
  await worker.run();
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});
```

</TabItem>
</Tabs>

## Workers are external to Temporal Server

In our tutorials, we show you how to run both the Temporal Server and one Worker on the same machine for local development.

However, a typical production Temporal deployment will have a **fleet** of Workers external to the main Temporal Server cluster. These can be independently managed by different developer teams, each registering their own sets of workflows and activities.

[![https://user-images.githubusercontent.com/6764957/113587567-8c9c9c00-9661-11eb-8614-576a68caa8f1.png](https://user-images.githubusercontent.com/6764957/113587567-8c9c9c00-9661-11eb-8614-576a68caa8f1.png)](https://docs.temporal.io/blog/workflow-engine-principles)

> Note: Temporal Server itself has [internal workers](https://docs.temporal.io/blog/workflow-engine-principles/#system-workflows-1910) for system workflows. But this is not visible to the developer.

> ## Workers can encrypt in transit and at rest

Author: ```suggestion
## Workers can be encrypted in transit and at rest
```

> However, a typical production Temporal deployment will have a **fleet** of Workers external to the main Temporal Server cluster. These can be independently managed by different developer teams, each registering their own sets of workflows and activities.
> 
> [![https://user-images.githubusercontent.com/6764957/113587567-8c9c9c00-9661-11eb-8614-576a68caa8f1.png](https://user-images.githubusercontent.com/6764957/113587567-8c9c9c00-9661-11eb-8614-576a68caa8f1.png)](https://docs.temporal.io/blog/workflow-engine-principles)
> 
> > Note: Temporal Server itself has [internal workers](https://docs.temporal.io/blog/workflow-engine-principles/#system-workflows-1910) for system workflows. But this is not visible to the developer.
> 
> ## Workers can be encrypted in transit and at rest
> 
> The external nature of workers works very well for data privacy concerns, because Temporal Server (including our managed Temporal Cloud version) doesn't run any Workflow or Activity code on its machines - it is solely responsible for orchestrating state transitions and dispatching messages to the next available Worker. 
> 
> While data transferred in the event histories is [secured by mTLS](https://docs.temporal.io/docs/server/security/#encryption-of-network-traffic), by default, it is still readable at rest in Temporal Server. Temporal offers a [Data Converter API](https://docs.temporal.io/docs/java/activities/#activity-interface) to solve this - you can customize the serialization of data going out of and coming back in to a Worker, with the net effect of guaranteeing that Temporal Server cannot read sensitive business data.

Author: ```suggestion
However, a typical production Temporal deployment will have a **fleet** of Workers external to the main Temporal Server cluster. 
These can be independently managed by different developer teams, each registering their own sets of workflows and activities.

[![https://user-images.githubusercontent.com/6764957/113587567-8c9c9c00-9661-11eb-8614-576a68caa8f1.png](https://user-images.githubusercontent.com/6764957/113587567-8c9c9c00-9661-11eb-8614-576a68caa8f1.png)](https://docs.temporal.io/blog/workflow-engine-principles)

> Note: Temporal Server itself has [internal workers](https://docs.temporal.io/blog/workflow-engine-principles/#system-workflows-1910) for system workflows. 
> But this is not visible to the developer.

## Workers can be encrypted in transit and at rest

The external nature of workers works very well for data privacy concerns, because Temporal Server (including our managed Temporal Cloud version) doesn't run any Workflow or Activity code on its machines. 
It is solely responsible for orchestrating state transitions and dispatching messages to the next available Worker. 

While data transferred in the event histories is [secured by mTLS](https://docs.temporal.io/docs/server/security/#encryption-of-network-traffic), by default, it is still readable at rest in Temporal Server. 

Temporal offers a [Data Converter API](https://docs.temporal.io/docs/java/activities/#activity-interface) to solve this - you can customize the serialization of data going out of and coming back in to a Worker, with the net effect of guaranteeing that Temporal Server cannot read sensitive business data.
```

sidebar_label: Workers
---

> A Worker is a service that hosts the [Workflow](/docs/concepts/workflows) and [Activity](/docs/concepts/activities) implementations. 

Author: ```suggestion
A Worker is a service that:

- hosts (registers) [Workflows](/docs/concepts/workflows) and [Activities](/docs/concepts/activities)
- listens to [Activity Task Queues](/docs/concepts/task-queues) by long polling

Workers should have access to the resources needed to perform the activities they implement, for example:

- network access for external API calls
- credentials for infrastructure provisioning
- specialized GPUs for machine learning tasks

Since workers are subscribed to task queues they can implement, it is straightforward to do [Activity Task Routing](/docs/concepts/activities/#activity-task-routing-through-task-queues). 

> Note: if you need to process stateful work sequentially on the same machine, the Go SDK also offers a [Sessions API](https://docs.temporal.io/docs/go/sessions/).
```

## Workers are external to Temporal Server

> In our tutorials, we show you how to run both the Temporal Server and Worker on the same machine for local development

Author: ```suggestion
In our tutorials, we show you how to run both the Temporal Server and Worker on the same machine for local development.
```

sidebar_label: Workers
---

> A Worker is a service that:

Reviewer: ```suggestion
A Worker is a service that does the following:
```

> - hosts (registers) [Workflows](/docs/concepts/workflows) and [Activities](/docs/concepts/activities)

Reviewer: ```suggestion
- Hosts executable [Workflow](/docs/concepts/workflows) and [Activity](/docs/concepts/activities) code.
```

> - listens to [Activity Task Queues](/docs/concepts/task-queues) by long polling

Reviewer: ```suggestion
- Listens to [Task Queues](/docs/concepts/task-queues) via long polling.
```
Not just Activity Task Queues right?

> Workers should have access to the resources needed to perform the activities they implement, for example:

Reviewer: ```suggestion
Workers must have access to any resources needed to execute the actions defined in Activities, such as the following:
```

> - network access for external API calls

Reviewer: ```suggestion
- Network access for external API calls.
```

> - credentials for infrastructure provisioning

Reviewer: ```suggestion
- Credentials for infrastructure provisioning.
```

> - specialized GPUs for machine learning tasks

Reviewer: ```suggestion
- Specialized GPUs for machine learning utilities.
```

> > Note: if you need to process stateful work sequentially on the same machine, the Go SDK also offers a [Sessions API](https://docs.temporal.io/docs/go/sessions/).

Reviewer: ```suggestion
:::note

If you need to process work sequentially on the same machine, the Go SDK also offers a [Sessions API](https://docs.temporal.io/docs/go/sessions/).

:::
```
stateful seems redundant in this sentence imo

See example Worker code for:

- [Go SDK](docs/go/workers)
- [Java SDK](docs/java/run-your-first-app-tutorial#the-worker)
- [PHP SDK](docs/php/workers)
- [Node SDK](docs/node/hello-world#worker)

> ## Workers are external to Temporal Server

Reviewer: ```suggestion
## Workers are external to the Temporal Server
```

In our tutorials, we show you how to run both the Temporal Server and one Worker on the same machine for local development.

However, a typical production Temporal deployment will have a **fleet** of Workers external to the main Temporal Server cluster. 

> These can be independently managed by different developer teams, each registering their own sets of workflows and activities.

Reviewer: ```suggestion
These can be independently managed by different developer teams, each registering their own sets of Workflows and Activities.
```

[![https://user-images.githubusercontent.com/6764957/113587567-8c9c9c00-9661-11eb-8614-576a68caa8f1.png](https://user-images.githubusercontent.com/6764957/113587567-8c9c9c00-9661-11eb-8614-576a68caa8f1.png)](https://docs.temporal.io/blog/workflow-engine-principles)

> > Note: Temporal Server itself has [internal workers](https://docs.temporal.io/blog/workflow-engine-principles/#system-workflows-1910) for system workflows. 

Reviewer: ```suggestion
:::note

Temporal Server itself has [internal workers](https://docs.temporal.io/blog/workflow-engine-principles/#system-workflows-1910) for system workflows.
But this is not visible to the developer.

:::
```

Not sure this is relevant to this concept at all...

> > But this is not visible to the developer.

Reviewer: ```suggestion

```

## Workers can be encrypted in transit and at rest

> The external nature of workers works very well for data privacy concerns, because Temporal Server (including our managed Temporal Cloud version) doesn't run any Workflow or Activity code on its machines. 

Reviewer: ```suggestion
The external nature of Workers works very well for data privacy concerns, because the Temporal Server (including our managed Temporal Cloud version) doesn't run any Workflow or Activity code on its machines. 
```

It is solely responsible for orchestrating state transitions and dispatching messages to the next available Worker. 

> While data transferred in the event histories is [secured by mTLS](https://docs.temporal.io/docs/server/security/#encryption-of-network-traffic), by default, it is still readable at rest in Temporal Server. 

Reviewer: ```suggestion
While data transferred in the event histories is [secured by mTLS](https://docs.temporal.io/docs/server/security/#encryption-of-network-traffic), by default, it is still readable at rest in the Temporal Server. 
```

> Temporal offers a [Data Converter API](https://docs.temporal.io/docs/java/activities/#activity-interface) to solve this - you can customize the serialization of data going out of and coming back in to a Worker, with the net effect of guaranteeing that Temporal Server cannot read sensitive business data.

Reviewer: ```suggestion
To solve this, Temporal SDKs offer a [Data Converter API](https://docs.temporal.io/docs/java/activities/#activity-interface) that you can use to customize the serialization of data going out of and coming back in to a Worker, with the net effect of guaranteeing that the Temporal Server cannot read sensitive business data.
```

sidebar_label: Workers
---

> import Tabs from '@theme/Tabs';
> import TabItem from '@theme/TabItem';
> 

Author: ```suggestion
```

sidebar_label: Workers
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

A Worker is a service that:

- hosts (registers) [Workflows](/docs/concepts/workflows) and [Activities](/docs/concepts/activities)
- listens to [Activity Task Queues](/docs/concepts/task-queues) by long polling

Workers should have access to the resources needed to perform the activities they implement, for example:

- network access for external API calls
- credentials for infrastructure provisioning
- specialized GPUs for machine learning tasks

> Since workers are subscribed to task queues they can implement, it is straightforward to do [Activity Task Routing](/docs/concepts/activities/#activity-task-routing-through-task-queues). 

Reviewer 2: I don't think this bit is necessary or helpful. Is task routing something very important that we want to highlight?

Author: alright i can take it out, Cully didnt like it either

```suggestion
```

> Note: if you need to process stateful work sequentially on the same machine, the Go SDK also offers a [Sessions API](https://docs.temporal.io/docs/go/sessions/).

> See example worker code for:

Reviewer 2: Uppercase "Worker"

Author: ```suggestion
See example Worker code for:
```

sidebar_label: Workers
---

A Worker is a service that does the following:

> - Hosts executable [Workflow](/docs/concepts/workflows) and [Activity](/docs/concepts/activities) code.

Reviewer 3: Should we add here that you can have separate workers for workflow and activities. And that a single worker can host both too?

Author: how about this

```suggestion
- Hosts executable [Workflow](/docs/concepts/workflows) and/or [Activity](/docs/concepts/activities) code. (Either can be hosted independently)
```

- Listens to [Task Queues](/docs/concepts/task-queues) via long polling.

> Workers must have access to any resources needed to execute the actions defined in Activities, such as the following:

Reviewer 3: This is a little confusing to me, does this mean i have to provide all my credentials to the worker when I register one? I'm not sure how to read / interpret these rules.

Author: just for the activities that you want to execute?

Reviewer 3: oh i see, yes you can pass parameters to activities when you register them like

worker.registerActivitiesImplementations(new GreetingActivitiesImpl("mySecretString));

I think in some cases users would use secrets or env vars instead of passing strings to activities..maybe it would be ok to put that as optional to not confuse users that they must do this ? just wondering

Author: i didn't say anything about *how* the credentials are passed - it can be as params, it can be as env vars :)

- Network access for external API calls.
- Credentials for infrastructure provisioning.
- Specialized GPUs for machine learning utilities.

:::note

If you need to process work sequentially on the same machine, the Go SDK also offers a [Sessions API](https://docs.temporal.io/docs/go/sessions/).

:::

See example Worker code for:

- [Go SDK](/docs/go/workers)
- [Java SDK](/docs/java/run-your-first-app-tutorial/#the-worker)
- [PHP SDK](/docs/php/workers)
- [Node SDK](/docs/node/hello-world/#worker)

## Workers are external to the Temporal Server

> In our tutorials, we show you how to run both the Temporal Server and one Worker on the same machine for local development.

Reviewer 3: which tutorials? (maybe we can add a link)

Author: hard to link, as there are different tutorials for every SDK
